python/ray/data/_internal/stats.py

```diff
@@ -258,15 +250,80 @@ def summary_string(
             else:
                 for i, metadata in stats_map.items():
                     self.stages["read"][i] = metadata[0]
+
+        stages_stats = []
+        is_substage = len(self.stages) > 1
+        for stage_name, metadata in self.stages.items():
+            stages_stats.append(
+                StageStatsSummary.from_block_metadata(
+                    metadata,
+                    self.time_total_s,
+                    stage_name,
+                    is_substage=is_substage,
+                )
+            )
+
+        iter_stats = IterStatsSummary(
+            self.iter_wait_s,
+            self.iter_get_s,
+            self.iter_next_batch_s,
+            self.iter_format_batch_s,
+            self.iter_user_s,
+            self.iter_total_s,
+        )
+        stats_summary_parents = []
+        if self.parents is not None:
+            stats_summary_parents = [p.to_summary() for p in self.parents]
+        return DatasetStatsSummary(
+            stages_stats,
+            iter_stats,
+            stats_summary_parents,
+            self.number,
+            self.dataset_uuid,
+            self.time_total_s,
+            self.base_name,
+            self.extra_metrics,
+        )
+
+
+@DeveloperAPI
+@dataclass
+class DatasetStatsSummary:
+    stages_stats: List["StageStatsSummary"]
+    iter_stats: "IterStatsSummary"
+    parents: List["DatasetStatsSummary"]
```

Comment:
IIUC, this assumes the execution contains only unary operators? If there are binary/n-ary operators, some stats may need to computed differently (e.g. walltime is not just simply summing up the walltime of all parents).

Reply:
@jianoaix I think this is an existing issue with the base `DatasetStats` object, right? https://github.com/ray-project/ray/blob/536737dc5211e81b82a3ef89b5f02a8e8eeccc5e/python/ray/data/_internal/stats.py#L187-L188

We definitely need to fix this for n-ary operators going forward (maybe once the execution plan is able to represent anything more than a linear chain of operators), but probably not in this PR.

Reply:
For the example of walltime in such cases, we currently sum up the max of its parents, which I believe is how we want to aggregate/handle the parents' stats:
https://github.com/ray-project/ray/blob/536737dc5211e81b82a3ef89b5f02a8e8eeccc5e/python/ray/data/_internal/stats.py#L361-L366

Do you think this is sufficient in covering the n-ary operator cases (and similarly with `get_total_cpu`/`get_max_heap_memory`)?

Reply:
Yeah agree that this is an existing issue with base `DatasetStats`. Agree to fix later, but not in this PR.

Reply:
It depends if the parent operators are running in parallel or synchronously, right? If they're running synchronously, then it's a `sum()`; if they're running in parallel, it's a `max()`, and this is all assuming that the operators started at the same time, which we could assume for parent blocks, but won't be able to assume for parent operators.

Reply:
Hmm actually, why is the parent calculation currently a `sum()` instead of a `max()`? Can't we assume that all parent blocks were launched at ~ the same time and executed in parallel, so the total parent wall-time is the max of each parent block?

Reply:
I think I see what you're saying @clarkzinzow, -- for the current `StageStatsSummary` object, we should be taking the max of its parents' wall times, and add that to the sum of current stages' wall times. So it should look like:
```
def get_total_wall_time(self) -> float: 
     parent_max_wall_time = max(p.get_total_wall_time() for p in self.parents)
     return parent_max_wall_time + sum( 
         getattr(ss, StageStatsSummary.WALL_TIME).get("max", 0) 
         for ss in self.stages_stats 
     ) 
```

Reply:
Actually wait, should I also be taking a max of the current stages' wall times here?
```
return parent_max_wall_time + max( 
         getattr(ss, StageStatsSummary.WALL_TIME).get("max", 0) 
         for ss in self.stages_stats 
     ) 
```

Reply:
The stages are executed sequentially, so I think that the `sum()` over the stage wall times is correct!

Although I'm wondering if we should just store start and end times in the stats for all blocks, and just recursively get the start time of the earliest block of the earliest root operator and subtract that from the end time of the latest block in this operator...

Reply:
Because the streaming executor is going to make these wall-time calculations difficult if we're just working with a DAG of operator wall-times, and there's also the delay in operator execution between operators that would be unaccounted for.

Reply:
Makes sense, maybe something we could address in a future PR? Since we have made several major refactors of the stats/summary classes, it may further nest the complexity.

Reply:
This comment was not about asking for fix, but exactly to check the assumption and start conversation/design regarding how to have a generic stats support. Feel free to proceed once other comments addressed.

Reply:
I think we should open an issue for this, tying it to the new execution model, namely making sure that the following is correct:
1. Wall-time calculation of n-ary operations.
2. Wall-time calculation under the streaming executor.